brain/detail/circuit.h

```diff
@@ -220,14 +187,313 @@ class Circuit::Impl
+                        std::to_string(expectedIdx) + "' got '" +
+                        std::to_string(nodeGroupIndex) + "'"));
+                }
+
+                const float x = attributeX[nodeGroupIndex];
+                const float y = attributeY[nodeGroupIndex];
+                const float z = attributeZ[nodeGroupIndex];
+
+                positions.push_back(Vector3f(x, y, z));
+
+                const float rX = rotationAngleX[nodeGroupIndex];
+                const float rY = rotationAngleY[nodeGroupIndex];
+                const float rZ = rotationAngleZ[nodeGroupIndex];
+
+                const float cX = std::cos(rX), cY = std::cos(rY),
+                            cZ = std::cos(rZ);
+
+                const float sX = std::sin(rX), sY = std::sin(rY),
+                            sZ = std::sin(rZ);
+
+                // These are the values given by multiplying the rotation
+                // matrices for R(X)R(Y)R(Z) i.e. extrinsic rotation around Z
+                // then Y then X
+                vmml::Matrix3f mtx;
+
+                mtx(0, 0) = cY * cZ;
+                mtx(0, 1) = -cY * sZ;
+                mtx(0, 2) = sY;
+                mtx(1, 0) = cZ * sX * sY + cX * sZ;
+                mtx(1, 1) = cX * cZ - sX * sY * sZ;
+                mtx(1, 2) = -cY * sX;
+                mtx(2, 0) = sX * sZ - cX * cZ * sY;
+                mtx(2, 1) = cZ * sX + cX * sY * sZ;
+                mtx(2, 2) = cX * cY;
+
+                rotations.push_back(Quaternionf(mtx));
+
+                expectedIdx++;
+            }
+        }
+    }
+    virtual ~SonataCircuit() {}
+    virtual size_t getNumNeurons() const
+    {
+        LBUNIMPLEMENTED;
+        return 0;
+    }
+
+    virtual GIDSet getGIDs(const std::string& /*target*/) const
+    {
+        LBUNIMPLEMENTED;
+        return GIDSet();
+    }
+
+    virtual Vector3fs getPositions(const GIDSet& gids) const
+    {
+        Vector3fs output;
+        for (auto gid : gids)
+            output.push_back(positions[gid]);
+        return output;
+    }
+    virtual size_ts getMTypes(const GIDSet& /*gids*/) const
+    {
+        LBUNIMPLEMENTED;
+        return size_ts();
+    }
+    virtual Strings getMorphologyNames() const
+    {
+        LBUNIMPLEMENTED;
+        return Strings();
+    }
+    virtual size_ts getETypes(const GIDSet& /*gids*/) const
+    {
+        LBUNIMPLEMENTED;
+        return size_ts();
+    }
+    virtual Strings getElectrophysiologyNames() const
+    {
+        LBUNIMPLEMENTED;
+        return Strings();
+    }
+    virtual Quaternionfs getRotations(const GIDSet& gids) const
+    {
+        Quaternionfs output;
+        for (auto gid : gids)
+            output.push_back(rotations[gid]);
+        return output;
+    }
+    virtual Strings getMorphologyNames(const GIDSet& /*gids*/) const
+    {
+        LBUNIMPLEMENTED;
+        return Strings();
+    }
+
+    virtual URI getMorphologyURI(const std::string& /*name*/) const
```

Comment:
For this function I've just realized that you can almost share the same implementation for all subclasses. You need to add a new virtual function getMorphologySource to replace the _morphologySource member variable in the original implementation.